dwave/preprocessing/src/presolveimpl.hpp

```diff
@@ -29,43 +31,206 @@ template <class Bias, class Index = int, class Assignment = double>
+                    overlap = true;
+                    constraint.mark_discrete(false);
+                    break;
+                }
+            }
+
+            if (overlap) {
+                discrete.erase(discrete.begin() + i);
+                continue;
+            }
+
+            ++i;
+        }
+    }
+
+    /// Remove the offsets from all constraints in the model.
+    void normalization_remove_offsets() {
+        for (auto& constraint : model_.constraints()) {
+            normalization_remove_offset(constraint);
+        }
+    }
+
+    /// Remove the offset from a cosntraint. E.g. `x + 1 <= 2` becomes `x <= 1`.
+    static void normalization_remove_offset(constraint_type& constraint) {
+        if (constraint.offset()) {
+            constraint.set_rhs(constraint.rhs() - constraint.offset());
+            constraint.set_offset(0);
+        }
+    }
+
+    /// Remove any self-loops (e.g. x^2) by adding a new variable and an equality constraint.
+    void normalization_remove_self_loops() {
+        std::unordered_map<index_type, index_type> mapping;
+
+        // Function to go through the objective/constraints and for each variable in a self-loop,
+        // we add a new variable it can interact with.
+        // We could break this out into a standalone method, but because all of  those methods
+        // would need to share a common mapping, there isn't really an opportunity for
+        // parallization. So let's keep it contained for now.
+        auto substitute = [&](expression_type& expression) {
+            std::size_t current_num_variables = expression.num_variables();
+            for (std::size_t i = 0; i < current_num_variables; ++i) {
+                index_type v = expression.variables()[i];
+
+                if (!expression.has_interaction(v, v)) continue;  // no self loop
+
+                auto it = mapping.find(v);
+                if (it == mapping.end()) {
+                    // we haven't seen this variable before
+                    it = mapping.emplace_hint(
+                            it, v,
+                            model_.add_variable(model_.vartype(v), model_.lower_bound(v),
+                                                model_.upper_bound(v)));
+                }
+
+                assert(it != mapping.end());
+
+                // Set the bias between v and the new variable
+                // We could spread the linear bias out between the two variables, but in that case
+                // we would probably want to do that in every expression for consistency and this
+                // is a lot easier. But it's something we could test in the future
+                expression.add_quadratic(v, it->second, expression.quadratic(v, v));
+                expression.remove_interaction(v, v);
+            }
+        };
+
+        // Actually apply the method
+        substitute(model_.objective());
+        for (auto& constraint : model_.constraints()) substitute(constraint);
```

Comment:
Addressed by https://github.com/dwavesystems/dwave-preprocessing/pull/101/commits/fda3ca4704bbbfdeaef2e7408379418f95e18deb. Though no doubt I managed to miss one or two.

Reply:
My personal preference is for brackets to always be used, even for one-liners.  However, I'm okay following a standard.  The code has a few of these inconsistencies; perhaps it makes sense for there to be a separate PR to do that?  At other jobs, we've set the standard in the formatter, thereby taking personal preferences or oversights out of the equation

Reply:
In this case I use clang format, but because this package is open sourced we don't want a formatter to be a requirement to all PRs because that raises the barrier for newer developers. We could possibly do it in CI, but haven't had the time to set that up. So for now, we accept a certain amount of inconsistency as the price for ease of contribution.

Reply:
Does look like we could change our [.clang-format](https://github.com/dwavesystems/dwave-preprocessing/blob/main/.clang-format) file to use [InsertBraces](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#insertbraces). That would probably be a separate PR though.